html2asketch/nodeToSketchLayers.js

```diff
@@ -181,8 +184,6 @@ export default function nodeToSketchLayers(node, options) {
       }
     }
 
-    style.addOpacity(opacity);
```

Comment:
let's put that into options and make enabled by default

Reply:
While trying to fix this I was wondering if the former implementation was even correct.
Currently, the opacity is set to the shape-group only, but not to any other layer, like text or SVG. So we get a transparent background and an opaque text on it.
I think we should either add the opacity to all layers, or remove it entirely, documenting that it is the container's responsibility to set the opacity.
What do you think?

Reply:
Good point - it should be added to all layers, but feel free to skip this fact in sake of this PR, I can handle it separately.

Reply:
I continued to think about this layers vs. group opacity thing, and I come to the conclusion that the problem is even worse. When adding opacity to each layer, the (outside coded) recursion will have to apply parent opacity to children as well, as the node children inside transparent parents will not issue any opacity value.
I think it is a better idea to leave out the opacity entirely in regards of the layers, and let the caller apply opacity to the parent. Any other approach results in a complex API, as far as I can see.
Or do you have a good idea as alternative?

Reply:
Hm, not sure if I fully understand.

1. For `nodeToSketchLayers` we should add opacity to all produced layers (with an option to turn opacity off for `nodeTreeToSketchGroup`)

1. For `nodeTreeToSketchGroup` we should set opacity on groups and not on the layers

<img width="1032" alt="screen shot 2018-04-18 at 12 04 54" src="https://user-images.githubusercontent.com/985504/38925748-ccba0ec4-4300-11e8-8d84-7ec1cde84aa9.png">

I don't see where the problem is 🤔

Reply:
Yes, that was the approach we discussed in the call. My issue is that I can hardly find an intact use-case for the default option with opacity layers.

As long as your DOM node does not have any children, everything is fine. The shape group and text nodes get the opacity, and it looks great.
But what if the DOM node has children (resp. parents have opacity)?

An example:
We have two nodes, A and B.
A has an opacity of 0.5, while B is opaque (no transparency).

The caller will be responsible for a recursion, and for A, nodeToSketchLayers will return layers with opacity 0.5, while for B, it will return opaque layers.
Now the caller wants to put all layers into one Sketch page.
If they flatten the layers, the node B layers will be opaque, which is wrong, as they are nested in the A node without opacity 0.5.
If he groups the nodes' layers hierarchically and puts the opacity into the grouping, they fail as well, as the node A layers' opacities are now in the way.

So basically my point is that setting a layer opacity does not work at all, unless the nodeToLayers function gets concise of its parents' opacity situation (which multipies up in the entire ancestry). This would have to be an API option, and I think this is not very clean.

Should we have a call? :-)

Reply:
As discussed in a call, we currently have three options:
- the current behavior, where only the current node's opacity is set into the layers
- not setting any layer opacity at all, which is required for the grouping function
- a solution where nodeToLayers collects the parents' opacities and multiplies them, so it is responsible for the parent opacity situation.

For this PR, I will implement options 1 and 2 (with 1 being the default).
In the future, the option is the best one when not grouping and should be another option (or a so-called fix for option 1).

Reply:
👍 I like the idea of making 3rd option a fix for option 1, so I went ahead and opened an issue for it - https://github.com/brainly/html-sketchapp/issues/84